include/envoy/buffer/buffer.h

```diff
@@ -194,6 +197,152 @@ class Instance {
+   * @param start supplies the buffer index to start copying from.
+   * @param Size how many bytes to read out of the buffer.
+   * @param Endianness specifies the byte order to use when decoding the integer.
+   * @details Size parameter: Some protocols have integer fields whose size in bytes won't match the
+   * size in bytes of C++'s integer types. Take a 3-byte integer field for example, which we want to
+   * represent as a 32-bit (4 bytes) integer. One option to deal with that situation is to read 4
+   * bytes from the buffer and ignore 1. There are a few problems with that solution, though.
+   *   * The first problem is buffer underflow: there may not be more than Size bytes available
+   * (say, last field in the payload), so that's an edge case to take into consideration.
+   *   * The second problem is draining the buffer after reading. With the above solution we cannot
+   *     read and discard in one go. We'd need to peek 4 bytes, ignore 1 and then drain 3. That not
+   *     only looks hacky since the sizes don't match, but also produces less terse code and
+   * requires the caller to propagate that logic to all call sites. Things complicate even further
+   * when endianness is taken into consideration: should the most or least-significant bytes be
+   * padded? Dealing with this situation requires a high level of care and attention to detail.
+   * Properly calculating which bytes to discard and how to displace the data is not only error
+   * prone, but also shifts to the caller a burden that could be solved in a much more generic,
+   * transparent and well tested manner.
+   *   * The last problem in the list is sign extension, which should be properly handled when
+   * reading signed types with negative values. To make matters easier, the optional Size parameter
+   * can be specified in those situations where there's a need to read less bytes than a C++'s
+   * integer size in bytes. For the most common case when one needs to read exactly as many bytes as
+   * the size of C++'s integer, this parameter can simply be omitted and it will be automatically
+   * deduced from the size of the type T
+   */
+  template <typename T, ByteOrder Endianness = ByteOrder::Host, size_t Size = sizeof(T)>
+  T peekInt(uint64_t start = 0) {
+    static_assert(Size <= sizeof(T), "requested size is bigger than integer being read");
+
+    if (length() < start + Size) {
+      throw EnvoyException("buffer underflow");
+    }
+
+    constexpr const auto displacement = Endianness == ByteOrder::BigEndian ? sizeof(T) - Size : 0;
+
+    auto result = static_cast<T>(0);
+    auto bytes = reinterpret_cast<char*>(std::addressof(result));
+    copyOut(start, Size, &bytes[displacement]);
+
+    constexpr const auto most_significant_read_byte =
+        Endianness == ByteOrder::BigEndian ? displacement : Size - 1;
+
+    constexpr const auto all_bits_enabled = static_cast<T>(~static_cast<T>(0));
+
+    const auto extension =
```

Comment:
To be extra clear, let's call this `sign_extension_bits`.

Reply:
👍